File: nonlinear_inversion_k_vector.m
clear all 
close all
%Scrip de prueba

%Entradas del algoritmo:
%   1)  y=f(x)
%   2)  [x_mi:x_max]
%   3)  n (numero de puntos a discretizar la funcion)

%Salida:
%   1)  N_roots (Numero de raices)
%   2)  X_e (?)


%Entradas (asi se puede separar mejor para luego hacer una funcion)
%*************************************************************************
%Preprocessing
%*************************************************************************

x_min=-5;
x_max=5;
n=10000; %Numero de muestras
y=zeros(n,1);
k_indice=1:n;% Vector de indices de x
K_vector=zeros(n,1);
x = x_min+((x_max-x_min)*(k_indice-1)/(n-1));   

y= sinc(x);
%y=-x.^2-4*x+4;

[s,I] = sort(y);

K_vector(1)=0;
K_vector(n)=n;

%Hacemos el k-vector para s
%*************************************************************************

y_min = s(1);
y_max = s(end); 

%Luego armamos la ecuacion de la recta z(x)=mx+q

n = length(y);

delta_eps = (n-1)*eps ;
m = (y_max-y_min+2*delta_eps)/(n-1);

q = y_min - m - delta_eps;

z = @(x)(m*x+q);
i=1:n;
z_i=z(i)';


for aux=2:length(y)-1
vec_aux=s<z_i(aux);
K_vector(aux)=sum(vec_aux);
end

%*************************************************************************

%Calculamos el delta
%*************************************************************************
delta=1/2*max(abs(y(2:n)-y(1:n-1)));
%*************************************************************************

%FIN Preprocesamiento




%Ahora vamos a calcular una busqueda para un y conocido(y=0) 
%*************************************************************************

y_conocido  = 0;

y_a=y_conocido-delta;

y_b=y_conocido+delta;




j_b=floor((y_a-q)/m);

j_t=ceil((y_b-q)/m);


k_start=K_vector(j_b)+1;

k_end=K_vector(j_t);

%Finalmente los elementos buscados y(i) \in [y_a;y_b] son los elementos y(I(k_busqueda))
%donde k_busqueda=k_start:k_end



k_busqueda=k_start:k_end;

y_busqueda=y(I(k_busqueda));
%Ahora si la cantidad de elementos de la base de datos es grande en promedio la
%cantidad de elementos que estan fuera del rango E0 =1 entonces esta en algunos
%de los dos extremos

if y_busqueda(1)<y_a
	y_busqueda(1)=[];%Removemos el elemento
end
if y_busqueda(end)>y_b
	y_busqueda(end)=[];%Removemos el elemento
end
y_busqueda;

X=x(I(k_busqueda));

Y=y(I(k_busqueda));

Roots=unique(round(X));
for indice=1:length(Roots)
    
    
    printf('Raiz:%d ---> %d\n',indice,Roots(indice));

end

%***********************************************************************

[X_s,J]=sort(X);

l=length(X_s);

File: nonlinear_inversion_k_vector.py
#!/usr/bin/env python
#*************************************************************************
#Imports
#*************************************************************************

import numpy as np
import matplotlib.pyplot as plt
from time import time

#Script de prueba
#
#Entradas del algoritmo:
#   1)  y=f(x)
#   2)  [x_mi:x_max]
#   3)  n (numero de puntos a discretizar la funcion)
#
#Salida:
#   1)  N_roots (Numero de raices)
#   2)  X_e (?)
#
#
#Entradas (asi se puede separar mejor para luego hacer una funcion)
#*************************************************************************
#Preprocessing
#*************************************************************************

x_min = -10. #El punto es importante ya que si no no es float
x_max = 10.

n=1000  #numero de muestras de nuestra base de datos

y=np.zeros(n,dtype=float)
x=np.zeros_like(y)
k_index=np.arange(n) #indices

x=x_min+((x_max-x_min)*(k_index-1)/(n-1))
#funcion a la que vamos a buscar las raices
y=np.sinc(x)

#*************************************************************************
#Realizamos el k-vector para y
#*************************************************************************

#prealocamos los vectores
s = np.zeros_like(y)
I = np.zeros_like(y)
k = np.zeros(n) #K-vector
y_busqueda = np.zeros_like(y)
vec_aux = np.zeros_like(y)

#epsilon es el numero de precision de la maquina
epsilon = np.finfo(np.float).eps
delta_epsilon = (n-1)*epsilon


I = np.argsort(y)

s = y[I]

y_min = s[0] #primer elemento
y_max = s[-1] #ultimo elemento

#Generamos la recta z(x)=mx+q

m = (y_max - y_min + 2 * delta_epsilon )/(n-1)
q = y_min - m - delta_epsilon
#funcion anonima para la recta
z = lambda x: (m*x+q)
i = np.arange(n)
z_discreto = z(i)

#Generamos el k-vector

k[0]=0
k[-1]=n

#TODO vectorizar :

for aux in xrange(1,n-1):
    vec_aux=s<z_discreto[aux]
    k[aux]=np.sum(vec_aux)

#Calculamos el delta
delta = .5*max(abs(y[1:n-1]-y[0:n-2]))#Recordar que 1/2 es 0 en  Python

#*************************************************************************
#Fin preprocesamiento
#*************************************************************************

#Ahora vamos a calcular una busqueda para un y conocido(y=0)
#*************************************************************************
y_conocido=0

y_a=y_conocido-delta
y_b=y_conocido+delta


j_b=np.floor((y_a-q)/m);

j_t=np.ceil((y_b-q)/m);

k_start = k[j_b]+1
k_end = k[j_t]

k_busqueda = np.arange(int(k_start),int(k_end))


y_busqueda = s[k_busqueda]

#Ahora si la cantidad de elementos de la base de datos es grande en promedio la
#cantidad de elementos que estan fuera del rango E0 =1 entonces esta en algunos
#de los dos extremos

if y_busqueda[0]<y_a:
    y_busqueda=np.delete(y_busqueda,0)

if y_busqueda[-1]>y_b:
    y_busqueda=np.delete(y_busqueda,len(y_busqueda)-1)

X=x[I[k_busqueda]]
Y=y[I[k_busqueda]]

Roots=np.unique(np.around(X))

for i in xrange(len(Roots)):

    print 'Raiz:%d ---> %d\n' %(i,Roots[i])



